src/http/server.cr

```diff
@@ -95,72 +100,187 @@ class HTTP::Server
+    tcp_server = bind(host, 0, reuse_port)
+    tcp_server.local_address.port
+  end
+
+  # Creates a `Socket::Server` listenting to *address* and adds it as an interface.
+  #
+  # The *address* can either be `host:port` or a Unix socket prefixed by `unix:`.
+  # This obviously works only on systems supporting Unix sockets.
   #
-  # For example you may let the system choose a port, then report it:
   # ```
-  # server = HTTP::Server.new(0) { }
-  # server.bind
-  # server.port # => 12345
+  # server = HTTP::Server.new { }
+  # server.bind "127.0.0.1:8080"
+  # server.bind "unix:/tmp/my-socket"
   # ```
-  def port
-    if server = @server
-      server.local_address.port.to_i
+  def bind(address : String, reuse_port : Bool = false) : Socket::Server
+    if address.starts_with?("unix:")
+      {% if flag?(:unix) %}
+      bind(Socket::UNIXAddress.new(address[5..-1])).as(Socket::Server)
+      {% else %}
+      raise NotImplementedError.new "system does not support Unix sockets"
+      {% end %}
     else
-      @port
+      host, _, port = address.partition(':')
+      bind(host, port.to_i, reuse_port).as(Socket::Server)
     end
   end
 
-  # Creates the underlying `TCPServer` if the doesn't already exist.
+  {% if flag?(:unix) %}
+  # Creates a `UNIXServer` bound to *address* and adds it as an interface.
   #
-  # You may set *reuse_port* to true to enable the `SO_REUSEPORT` socket option,
-  # which allows multiple processes to bind to the same port.
-  def bind(reuse_port = false)
-    @server ||= TCPServer.new(@host, @port, reuse_port: reuse_port)
+  # ```
+  # server = HTTP::Server.new { }
+  # server.bind Socket::UNIXAddress.new("/tmp/my-socket")
+  # ```
+  def bind(address : Socket::UNIXAddress) : UNIXServer
+    bind UNIXServer.new(address.path)
   end
+  {% end %}
 
-  # Starts the server. Blocks until the server is closed.
+  # Creates a `TCPServer` bound to *address* and adds it as an interface.
   #
-  # See `#bind` for details on the *reuse_port* argument.
-  def listen(reuse_port = false)
-    server = bind(reuse_port)
+  # ```
+  # server = HTTP::Server.new { }
+  # server.bind Socket::IPAddress.new("127.0.0.1", 12345)
+  # ```
+  def bind(address : Socket::IPAddress) : TCPServer
+    bind address.host, address.port, reuse_port
+  end
+
+  # Adds a `Socket::Server` *interface* to this server.
+  def bind(interface)
+    @interfaces << interface
+
+    interface
+  end
+
+  # Binds to the specified arguments and starts the server.
+  #
+  # See `#bind` and `#listen` for details
+  def listen(*args, **kwargs)
+    bind *args, **kwargs
+
+    listen
+  end
+
+  # Starts the server. Blocks until the server is closed.
+  def listen
+    raise "can't start server without interfaces" if @interfaces.empty?
+
     until @wants_close
-      spawn handle_client(server.accept?)
+      @interfaces.each do |interface|
+        spawn handle_client(interface.accept?)
```

Comment:
This is completely wrong. It will round-robin each interface accepting one connection from each interface in order, instead of which interface is ready to accept. You need to spawn a fiber for each interface.

Reply:
Darn. I'll put my shame hat on and hide in the basement... :(
The worst of it is that I had a relevant spec for this which somehow got lost during rebasing.